### PR TITLE
Egl context fixes

### PR DESCRIPTION
**Pull request by xen2**
Various Egl bugfixes:
- Egl & Cocoa: Don't change context when creating/deleting them
- Egl: Unset context when calling MakeCurrent(null)
- Egl: Fix context sharing

### CODE REVIEW

**Comment by cra0zy** [comment]
Looks good to me.

cc. @thefiddler @amulware @varon @wcdeich4

**Comment by varon** [comment]
@xen2 Thanks for the PR! @cra0zy Thanks for the heads up.

I'll take a careful look through it in a bit.

**Comment by wcdeich4** [comment]
@xen2 what systems did you test this on?

File: Source/OpenTK/Platform/Egl/EglContext.cs [hunk 1]
```diff
@@ -97,7 +95,7 @@ abstract class EglContext : EmbeddedGraphicsContext
                 window.CreateWindowSurface(config);
 
             int[] attrib_list = new int[] { Egl.CONTEXT_CLIENT_VERSION, major, Egl.NONE };
-            HandleAsEGLContext = Egl.CreateContext(window.Display, config, shared != null ? shared.HandleAsEGLContext : IntPtr.Zero, attrib_list);
+            HandleAsEGLContext = Egl.CreateContext(window.Display, config, sharedContext != null ? (sharedContext as IGraphicsContextInternal).Context.Handle : IntPtr.Zero, attrib_list);
```
Comment (varon) [inline Source/OpenTK/Platform/Egl/EglContext.cs]: The cast on SharedContext violates [Liskov](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

As a quick fix for robustness, could you perform another null-check after the cast? The alternative path handling can be really simple, use your best judgement for that.

File: Source/OpenTK/Platform/Egl/EglContext.cs [hunk 2]
```diff
@@ -97,9 +95,7 @@ abstract class EglContext : EmbeddedGraphicsContext
                 window.CreateWindowSurface(config);
 
             int[] attrib_list = new int[] { Egl.CONTEXT_CLIENT_VERSION, major, Egl.NONE };
-            HandleAsEGLContext = Egl.CreateContext(window.Display, config, shared != null ? shared.HandleAsEGLContext : IntPtr.Zero, attrib_list);
-
-            MakeCurrent(window);
+            HandleAsEGLContext = Egl.CreateContext(window.Display, config, sharedContext != null ? (sharedContext as IGraphicsContextInternal).Context.Handle : IntPtr.Zero, attrib_list);
```
Comment (varon) [inline Source/OpenTK/Platform/Egl/EglContext.cs]: The cast on SharedContext violates [Liskov](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

As a quick fix for robustness, could you perform another null-check after the cast? The alternative path handling can be really simple, use your best judgement for that.

**Comment by varon** [comment]
Aside from that one cast, I can't see anything severe here. Thanks for the nice work, @xen2.

File: Source/OpenTK/Platform/Egl/EglContext.cs (hunk 2 above)
Comment (xen2) [inline Source/OpenTK/Platform/Egl/EglContext.cs]: Sure, I will improve it.
So far, I simply copy-pasted the existing code done for shared context in most other implementations (Windows, Mac, etc...); should I fix those too?

**Comment by xen2** [comment]
@wcdeich4 Android and iOS; I didn't test on Mac and after rechecking it again, I thought a `MakeContext()` was done in `CocoaNativeWindow` but it doesn't seem to be the case, so I probably need to double-check that again.

File: Source/OpenTK/Platform/Egl/EglContext.cs (hunk 2 above)
Comment (varon) [inline Source/OpenTK/Platform/Egl/EglContext.cs]: Ah, good point. As long as it's no-worse than our current ones, It's okay as-is.

If you want to, you can fix them, but it shouldn't be necessary for us to merge this PR.

**Comment by varon** [comment]
I can run a quick test on a mac here.

**Comment by cra0zy** [comment]
Thanks @xen2, Merging!!!

**Pull request merged**